Add thprac support

Bai-Chiang opened the pull request:
- Add [thprac](https://github.com/touhouworldcup/thprac) support with additional `-p /path/to/thprac.exe` option.

It will launch the game first then wait 5sec launch thprac with the same WINEPREFIX, so thprac would detect running game and automatically attach to it.

I used suggest command from this comment https://github.com/touhouworldcup/thprac/issues/129#issuecomment-1475398643
and some @redirectto scripts https://github.com/redirectto/thprac-steam-proton-wrapper

I tested with native Steam installation (on Steam Deck) and flatpak version of Steam.

- Use $XDG_DATA_HOME for default location
- Also kill custom.exe

Bai-Chiang committed:
Add thprac support.

Co-authored-by: redirectto <er.redirectto@gmail.com>

Bai-Chiang committed:
Update README.md

nerusuki commented on thcrap_proton:
```diff
@@ -91,6 +92,17 @@ echo "---------------------------------------------------------"
+    # wait 5s and launch thprac so it could auto attach to the game
+    sleep 5
+    PROTON_CMD=$( echo "$COMMAND" | sed -n "s/^.*-- \(.*\) waitforexitandrun.*$/\\1/p" )
+    RUN_THPRAC="WINEPREFIX='$STEAM_COMPAT_DATA_PATH/pfx/' $PROTON_CMD run '$THPRAC_EXE'"
```
Setting the ```WINEPREFIX``` is not necessary, since proton already uses the ```STEAM_COMPAT_DATA_PATH``` for this.

Bai-Chiang replied:
Thanks, I didn't know this. Yes, removing `WINEPREFIX` still works.

tactikauan commented:
Hi. Thanks for the PR. It wold be great if we could automatically install thprac like we already do with thcrap, so the user doesn't have to specify the path to the executable. There doesn't appear to be a single download link available to the latest version, but this should do the trick:

    VERSION=$(curl -si https://github.com/touhouworldcup/thprac/releases/latest | grep location: | sed 's@.*/@@' | tr -d '[:space:]')
    curl -L -f -O "https://github.com/touhouworldcup/thprac/releases/download/$VERSION/thprac.$VERSION.exe" --output-dir "$THPRAC_FOLDER"

Tough, if you prefer, I can merge this as-is and implement that at a later point.

nerusuki commented on README.md:
```diff
@@ -43,27 +43,38 @@ If you have gone through the manual installation, you have the opportunity to ch
-So, if I wanted to run the game with vpatch and Brazilian Portuguese translations, the command would look like this:
-       
-    thcrap_proton -v -c pt-br.js -- %command%
+To enable thprac for that game, use the `-p` flag to specify thprac executable.
+Upon launching thprac will ask whether to apply to the ongoing game, if running game
+in full screen mode you may need to minimize the game to see the dialog box.
+To [change thprac language](https://github.com/touhouworldcup/thprac?tab=readme-ov-file#how-do-i-switch-language),
+first temporarily change the launch option to _Launch Controller Layout Tool_.
+The drop down menu is located above where you change launch command.
+This will start the thprac launcher, and you can change lauguage under settings.
+Or you can set `LANG` environment variable to `en_US.utf8` for example.
+See below how to set environment variables.
+The launcher setting will take precedence over environment variable.
+
+So, if I wanted to run the game with vpatch and thprac and Brazilian Portuguese translations, the command would look like this:
+
+    thcrap_proton -v -c pt-br.js -p /path/to/thprac.vx.x.x.x.exe -- %command%
```
I think it's better to keep the original example intact for people just trying to get the most common use-case working, and then introduce the thprac-specific instructions with another example.

redirectto commented:
I'm glad you are taking a look at my script, the main flaw I have with it is that I cannot make it work properly with non-steam games, aside from running thcrap on the same folder as the games.
Aside from the flag that lauches thprac (-p), I added two flags, one to launch custom.exe (-k) and a second one to disable thcrap (-d).
I also added dectection to Uwabami Breakers (alcostg)
I tried the command @Bai-Chiang posted to see if it works with non-steam games but it doesn't, still I'll take it into account when making scripts that launch multiple apps on the same prefix.

Bai-Chiang committed:
Auto download thprac

Co-authored-by: Kauan Decarli <kauandecarli@gmail.com>

nerusuki commented on thcrap_proton:
```diff
@@ -71,6 +73,28 @@ if [ ! -f "$THCRAP_FOLDER/config/$THCRAP_CONFIG" ]; then
+            USE_THPRAC=0
+        fi
+    fi
+fi
```
This will fail when there's no Internet connection. Since it's just an executable and doesn't carry around user config files, we could just install/update it silently. Let's also put a timeout in the first curl to avoid stalling the game launch for too long.

```suggestion
if [ "$USE_THPRAC" == 1 ]; then
    THPRAC_VERSION=$(LD_LIBRARY_PATH=/usr/lib64:/usr/lib curl --connect-timeout 5 -si https://github.com/touhouworldcup/thprac/releases/latest | grep location: | sed 's@.*/@@' | tr -d '[:space:]')
    if [[ ! -z "$THPRAC_VERSION" && ! -f "$THPRAC_FOLDER/thprac.$THPRAC_VERSION.exe" ]]; then
        mkdir -p "$THPRAC_FOLDER"
        if [ ! -w "$THPRAC_FOLDER" ]; then
            ${ZENITY} --error --text="The directory \"$THPRAC_FOLDER\" could not be created or is not writable. Check your permissions and try again."
            exit
        fi
        LD_LIBRARY_PATH=/usr/lib64:/usr/lib curl -L -f "https://github.com/touhouworldcup/thprac/releases/download/$THPRAC_VERSION/thprac.$THPRAC_VERSION.exe" -o "$THPRAC_FOLDER/thprac.$THPRAC_VERSION.exe" | ${ZENITY} --progress --pulsate --auto-close --text="Downloading thprac.$THPRAC_VERSION.exe"
        if [ $? != 0 ] || [ ! -f "$THPRAC_FOLDER/thprac.$THPRAC_VERSION.exe" ]; then
            # Only abort if there's no version currently installed
            if [ ! -f "$THPRAC_FOLDER/thprac.exe" ]; then
                ${ZENITY} --error --text="Download failed. Aborting"
                exit
            fi
        else
            # Create a symlink to latest version
            ln -srf "$THPRAC_FOLDER/thprac.$THPRAC_VERSION.exe" "$THPRAC_FOLDER/thprac.exe"
        fi
    fi
fi
```

nerusuki commented on thcrap_proton:
```diff
@@ -71,6 +75,40 @@ if [ ! -f "$THCRAP_FOLDER/config/$THCRAP_CONFIG" ]; then
+                exit
+            fi
+        else
+            # Remove old versions, only keep 5 latest versions
+            readarray -d '' THPRAC_OLD_VERSIONS < <(find "$THPRAC_FOLDER" -maxdepth 1 -regextype egrep -iregex '.*/thprac.v([0-9]|\.)*.exe' -print0 | sort -zVr | tail -z -n +6)
+            if (( ${#THPRAC_OLD_VERSIONS[@]} )); then
+                # THPRAC_OLD_VERSIONS not empty
+                rm "${THPRAC_OLD_VERSIONS[@]}"
+            fi
```
For the sake of simplicity I'd prefer to not even bother with the older versions and just leave everything there.

Bai-Chiang replied:
I still prefer it could automatically cleanup unused versions.

What about only keep latest version with this single command
```bash
find "$THPRAC_FOLDER" -maxdepth 1 \! -name "thprac.$THPRAC_VERSION.exe" -exec rm {} \;`
```

or replace _if_ expression with
```
(( ${#THPRAC_OLD_VERSIONS[@]} )) && rm "${THPRAC_OLD_VERSIONS[@]}"
```
it just **looks** less cluttered though.

nerusuki replied:
Keeping only the latest version seems the most solid approach. Though we better filter exactly what we want to delete, otherwise we might end up accidentally erasing user files if they set `THPRAC_FOLDER` incorrectly, e.g., to their home directory.

    find "$THPRAC_FOLDER" -maxdepth 1 -name "thprac.*.exe" \! -name "thprac.$THPRAC_VERSION.exe" -exec rm {} \;

nerusuki commented on thcrap_proton:
```diff
@@ -71,6 +75,40 @@ if [ ! -f "$THCRAP_FOLDER/config/$THCRAP_CONFIG" ]; then
+                rm "${THPRAC_OLD_VERSIONS[@]}"
+            fi
+            # Create a symlink to latest version
+            # You can manually change launch version by pointing the symlink to old version.
```
Is there any reason to run an old version? (I genuinely ask since I never used the tool)

Bai-Chiang replied:
Well, it's also my first time using thprac. I thought this could roll back, in case bug in latest release. Probably it's not a good approach. I could remove this comment if you prefer.

nerusuki replied:
Yeah I don't think it makes sense.

nerusuki commented on thcrap_proton:
```diff
@@ -71,6 +75,40 @@ if [ ! -f "$THCRAP_FOLDER/config/$THCRAP_CONFIG" ]; then
     fi
 fi
 
+if [ "$USE_THPRAC" == 1 ]; then
+    THPRAC_VERSION=$(LD_LIBRARY_PATH=/usr/lib64:/usr/lib curl --connect-timeout 5 -si https://github.com/touhouworldcup/thprac/releases/latest | grep location: | sed 's@.*/@@' | tr -d '[:space:]')
+    if [[ -n "$THPRAC_VERSION" && ! -f "$THPRAC_FOLDER/thprac.$THPRAC_VERSION.exe" ]]; then
+        # Install latest version
+        mkdir -p "$THPRAC_FOLDER"
+        if [ ! -w "$THPRAC_FOLDER" ]; then
+            ${ZENITY} --error --text="The directory \"$THPRAC_FOLDER\" could not be created or is not writable. Check your permissions and try again."
+            exit
+        fi
+        LD_LIBRARY_PATH=/usr/lib64:/usr/lib curl -L -f "https://github.com/touhouworldcup/thprac/releases/download/$THPRAC_VERSION/thprac.$THPRAC_VERSION.exe" -o "$THPRAC_FOLDER/thprac.$THPRAC_VERSION.exe" | ${ZENITY} --progress --pulsate --auto-close --text="Downloading \"thprac.$THPRAC_VERSION.exe\""
+        if [ $? != 0 ]; then
+            # Only abort if there's no version currently installed
+            if [ ! -f "$THPRAC_FOLDER/thprac.exe" ]; then
+                ${ZENITY} --error --text="Download failed. Aborting"
+                exit
+            fi
+        else
+            # Remove old versions, only keep 5 latest versions
+            readarray -d '' THPRAC_OLD_VERSIONS < <(find "$THPRAC_FOLDER" -maxdepth 1 -regextype egrep -iregex '.*/thprac.v([0-9]|\.)*.exe' -print0 | sort -zVr | tail -z -n +6)
+            if (( ${#THPRAC_OLD_VERSIONS[@]} )); then
+                # THPRAC_OLD_VERSIONS not empty
+                rm "${THPRAC_OLD_VERSIONS[@]}"
+            fi
+            # Create a symlink to latest version
+            # You can manually change launch version by pointing the symlink to old version.
+            ln -srf "$THPRAC_FOLDER/thprac.$THPRAC_VERSION.exe" "$THPRAC_FOLDER/thprac.exe"
+        fi
+    elif [[ -z "$THPRAC_VERSION" && ! -f "$THPRAC_FOLDER/thprac.exe" ]]; then
+        # no connection and no installed version
+        ${ZENITY} --error --text="Download failed. Aborting"
+        exit
+    fi
+fi
```
We can simplify things by checking for the installation only once at the end. Let's also make the message more clear like "thprac installation failed. Aborting".

Bai-Chiang replied:
Sounds good.

nerusuki commented on README.md:
```diff
@@ -26,6 +26,10 @@ Also, with Steam Play/Proton, it is expected that you can run your games without
 
     flatpak install flathub com.valvesoftware.Steam.Utility.thcrap_steam_proton_wrapper
 
+**Note: For flatpak version of Steam, by default `/home/$USER` or `$HOME` directory is mapped to `/home/$USER/.var/app/com.valvesoftware.Steam`.
+A dialog box will ask whether to install thcrap to `/home/$USER/.local/share/thcrap`,
+but the real location is `/home/$USER/.var/app/com.valvesoftware.Steam/.local/share/thcrap`**
+
```
I'd like to keep the installation guide straightforward. The user probably doesn't care where things end up being installed.

Bai-Chiang replied:
I think the best solution would be using `$XDG_DATA_HOME`[1] instead of `/home/$USER/.local/share` for default location like this
```bash
[ -z "$XDG_DATA_HOME" ] && XDG_DATA_HOME="$HOME/.local/share"
THCRAP_FOLDER="$XDG_DATA_HOME/thcrap"
```

flatpak will overwrite[2] this env variable, so it will show correct path.
Then we don't need those extra notes, otherwise it will cause some confusion.

[1] https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html#variables
[2] https://docs.flatpak.org/en/latest/sandbox-permissions-reference.html#f6

nerusuki replied:
Indeed, using `$XDG_DATA_HOME` shows the correct path inside the sandbox. It also seems more correct overall. Nice solution!

Bai-Chiang committed:
Use $XDG_DATA_HOME for default location.

This handles better for flatpak installation which should show correct
installation location.

Bai-Chiang committed:
Don't start thprac when launching config tool

Bai-Chiang committed:
Also kill config executable

nerusuki commented on thcrap_proton:
```diff
@@ -90,10 +121,20 @@ echo "$COMMAND"
 # Get PID while ensuring only the actual Touhou executable is chosen
 while [ -z "$TH_PID" ]; do # Loop until a matching process is found
-    TH_PID=$(grep -Er "$TH_REGEX" /proc/*/stat | sed "s/.*:\([0-9]\+\).*/\\1/")
+    TH_PID=$(grep -Er "($TH_REGEX\.exe|custom\.exe)" /proc/*/stat | sed "s/.*:\([0-9]\+\).*/\\1/")
```
This will not work for EoSD because the executable name in `/proc/[pid]/stat` is limited to 15 bytes, and `東方紅魔郷.exe` is 19 bytes, so it gets truncated to just `東方紅魔郷`.

nerusuki replied:
> Okay, I didn't know that. Does `"($TH_REGEX|custom\.exe)"` works?

`"($TH_REGEX|custom\.exe)"` should work fine.

Bai-Chiang commented:
Okay, I didn't know that. Does `"($TH_REGEX|custom\.exe)"` works?

On March 24, 2024 8:27:12 AM EDT, Kauan Decarli ***@***.***> wrote:
>@tactikauan commented on this pull request.
>
>
>
>>  
> # Get PID while ensuring only the actual Touhou executable is chosen
> while [ -z "$TH_PID" ]; do # Loop until a matching process is found
>-    TH_PID=$(grep -Er "$TH_REGEX" /proc/*/stat | sed "s/.*:\([0-9]\+\).*/\\1/")
>+    TH_PID=$(grep -Er "($TH_REGEX\.exe|custom\.exe)" /proc/*/stat | sed "s/.*:\([0-9]\+\).*/\\1/")
>
>This will not work for EoSD because the executable name in `/proc/[pid]/stat` is limited to 15 bytes, and `東方紅魔郷.exe` is 19 bytes, so it gets truncated to just `東方紅魔郷`.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/tactikauan/thcrap-steam-proton-wrapper/pull/9#pullrequestreview-1956539164
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>

tactikauan commented:
Other than the comment above looks good. I think it's a good idea to reduce the number of commits for better readability. Suggestion:

- Add thprac support
- Use $XDG_DATA_HOME for default location
- Also kill custom.exe

Also, please add @redirectto as a co-author to the relevant commit(s).

Bai-Chiang commented:
Added @redirectto to co-author, the only new commit is https://github.com/tactikauan/thcrap-steam-proton-wrapper/pull/9/commits/e041601d8d6ae4b34ea2d273c71826b0c5133dc4

tactikauan commented:
Thanks. I squashed and reorganized some commits for a cleaner history.
